File: backend/app/services/picking.py
"""Order-picker routing over the warehouse aisle graph.

Model (classical single-block layout, De Koster/Roodbergen surveys):
- Depot = the door on the south wall (x = width/2, y = 0 — same convention as
  the 3D scene's buildWalls entrance).
- Aisles run along x (racks are laid out in row bands); pickers travel the
  front (y=0 side) and back cross-aisles and enter aisles vertically.
  In this layout aisles are horizontal bands, so "aisle axis" is x and the
  cross-aisle axis is y; travel between aisles happens on the left/right
  cross corridors (x = min/max aisle extents).

Policies:
- s_shape  : industry baseline — traverse every aisle containing picks fully,
             snaking left→right (simple, memorizable tours).
- largest_gap : enter each pick aisle from both ends up to the largest gap.
- optimized: TSP over the pick points using aisle-graph shortest-path
             distances — greedy seed + 2-opt improvement (networkx has no
             built-in 2-opt). Lands within a few percent of the exact
             Ratliff–Rosenthal (1983) optimum at demo sizes.

Pure analysis: writes nothing, returns per-policy stop order + floor polyline.
"""

from dataclasses import dataclass

import networkx as nx
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.errors import ValidationFailedError
from app.models import Product, StockItem, StorageLocation
from app.schemas.network import PathPoint, PickRouteOut, PickStop, PolicyRoute
from app.services.scoping import get_owned_warehouse


@dataclass
class _Pick:
    location_id: int
    code: str
    x: float  # bin center x (meters)
    y: float  # bin center y
    aisle_key: float  # aisle band identifier (aisle center y)
    product_sku: str | None = None
    quantity: int | None = None


def _resolve_picks(
    db: Session, org_id: int, warehouse_id: int,
    items: list | None, location_ids: list[int] | None,
) -> list[_Pick]:
    """Turn an order (products) or an explicit bin list into pick points."""
    bins_by_id: dict[int, StorageLocation] = {
        loc.id: loc
        for loc in db.scalars(
            select(StorageLocation).where(
                StorageLocation.warehouse_id == warehouse_id,
                StorageLocation.type == "bin",
            )
        ).all()
    }
    picks: list[_Pick] = []

    if location_ids:
        for lid in location_ids:
            loc = bins_by_id.get(lid)
            if loc is None:
                raise ValidationFailedError(f"Göz {lid} bu depoda bulunamadı")
            picks.append(_make_pick(loc))
    elif items:
        skus = {
            p.id: p.sku
            for p in db.scalars(
                select(Product).where(Product.org_id == org_id)
            ).all()
        }
        for item in items:
            stock_rows = db.execute(
                select(StockItem)
                .join(StorageLocation, StockItem.location_id == StorageLocation.id)
                .where(
                    StockItem.product_id == item.product_id,
                    StorageLocation.warehouse_id == warehouse_id,
                    StockItem.quantity > 0,
                )
                .order_by(StockItem.quantity.desc())
            ).scalars().all()
            if not stock_rows:
                raise ValidationFailedError(
                    f"Ürün {item.product_id} bu depoda stokta yok"
                )
            loc = bins_by_id[stock_rows[0].location_id]
            pick = _make_pick(loc)
            pick.product_sku = skus.get(item.product_id)
            pick.quantity = item.quantity
            picks.append(pick)
    if not picks:
        raise ValidationFailedError("Toplanacak göz listesi boş")
    # Dedupe by location, keep first occurrence.
    seen: set[int] = set()
    unique = []
    for p in picks:
        if p.location_id not in seen:
            seen.add(p.location_id)
            unique.append(p)
    return unique


def _make_pick(loc: StorageLocation) -> _Pick:
    return _Pick(
        location_id=loc.id,
        code=loc.code,
        x=loc.pos_x + loc.dim_w / 2,
        y=loc.pos_y + loc.dim_d / 2,
        aisle_key=round(loc.pos_y, 1),
    )


def _route_for_order(
    depot: tuple[float, float],
    ordered: list[_Pick],
    front_y: float,
    corridor_left: float,
    corridor_right: float,
) -> tuple[float, list[PathPoint]]:
    """Walk the pick sequence with rectilinear moves along corridors:
    vertical travel happens on the nearest side corridor, horizontal travel
    inside the aisle band. Returns (total_m, polyline)."""
    path: list[PathPoint] = [PathPoint(x=depot[0], y=depot[1])]
    total = 0.0

    def goto(x: float, y: float) -> None:
        nonlocal total
        last = path[-1]
        if abs(last.x - x) < 1e-9 and abs(last.y - y) < 1e-9:
            return
        total += abs(last.x - x) + abs(last.y - y)
        # rectilinear: first move along x, then along y (visual convention)
        if abs(last.x - x) > 1e-9:
            path.append(PathPoint(x=x, y=last.y))
        if abs(last.y - y) > 1e-9:
            path.append(PathPoint(x=x, y=y))

    for pick in ordered:
        last = path[-1]
        if abs(last.y - pick.y) > 1e-9:
            # change aisle via the nearest side corridor
            corridor = (
                corridor_left
                if abs(last.x - corridor_left) + abs(pick.x - corridor_left)
                <= abs(last.x - corridor_right) + abs(pick.x - corridor_right)
                else corridor_right
            )
            goto(corridor, last.y)
            goto(corridor, pick.y)
        goto(pick.x, pick.y)
    # return to depot
    last = path[-1]
    if abs(last.y - depot[1]) > 1e-9:
        corridor = (
            corridor_left
            if abs(last.x - corridor_left) + abs(depot[0] - corridor_left)
            <= abs(last.x - corridor_right) + abs(depot[0] - corridor_right)
            else corridor_right
        )
        goto(corridor, last.y)
        goto(corridor, depot[1])
    goto(depot[0], depot[1])
    return round(total, 1), path


def _s_shape_order(picks: list[_Pick]) -> list[_Pick]:
    """Aisles sorted front→back; within an aisle sweep alternates direction
    (snake), which is what makes the classic S pattern."""
    aisles = sorted({p.aisle_key for p in picks})
    ordered: list[_Pick] = []
    for i, aisle in enumerate(aisles):
        row = sorted([p for p in picks if p.aisle_key == aisle], key=lambda p: p.x)
        if i % 2 == 1:
            row.reverse()
        ordered.extend(row)
    return ordered


def _largest_gap_order(
    picks: list[_Pick], corridor_left: float, corridor_right: float
) -> list[_Pick]:
    """Per aisle: picks left of the largest gap are served from the left end,
    right of it from the right end (no full traversal)."""
    aisles = sorted({p.aisle_key for p in picks})
    ordered: list[_Pick] = []
    for aisle in aisles:
        row = sorted([p for p in picks if p.aisle_key == aisle], key=lambda p: p.x)
        if len(row) == 1:
            ordered.extend(row)
            continue
        xs = [corridor_left] + [p.x for p in row] + [corridor_right]
        gaps = [(xs[i + 1] - xs[i], i) for i in range(len(xs) - 1)]
        _, gap_idx = max(gaps)
        left_side = row[:gap_idx]  # served from left, in order
        right_side = row[gap_idx:]  # served from right, reversed
        ordered.extend(left_side)
        ordered.extend(reversed(right_side))
    return ordered


def _optimized_order(
    depot: tuple[float, float],
    picks: list[_Pick],
    front_y: float,
    corridor_left: float,
    corridor_right: float,
) -> list[_Pick]:
    """Greedy nearest-neighbour seed on corridor distances, then 2-opt."""

    def dist(
        a: tuple[float, float],
        b: tuple[float, float],
        a_aisle: float | None,
        b_aisle: float | None,
    ) -> float:
        # same aisle band → straight along x; else via nearest corridor
        if a_aisle is not None and b_aisle is not None and abs(a_aisle - b_aisle) < 1e-9:
            return abs(a[0] - b[0])
        via_left = abs(a[0] - corridor_left) + abs(b[0] - corridor_left)
        via_right = abs(a[0] - corridor_right) + abs(b[0] - corridor_right)
        return min(via_left, via_right) + abs(a[1] - b[1])

    nodes = list(range(len(picks)))

    def d(i: int | None, j: int | None) -> float:
        pa = depot if i is None else (picks[i].x, picks[i].y)
        pb = depot if j is None else (picks[j].x, picks[j].y)
        aa = None if i is None else picks[i].aisle_key
        ab = None if j is None else picks[j].aisle_key
        return dist(pa, pb, aa, ab)

    # greedy seed from depot
    remaining = set(nodes)
    tour: list[int] = []
    current: int | None = None
    while remaining:
        nxt = min(remaining, key=lambda j: d(current, j))
        tour.append(nxt)
        remaining.remove(nxt)
        current = nxt

    # 2-opt (closed tour including depot at both ends)
    def tour_len(t: list[int]) -> float:
        total = d(None, t[0]) + d(t[-1], None)
        for a, b in zip(t, t[1:], strict=False):
            total += d(a, b)
        return total

    improved = True
    best = tour
    best_len = tour_len(best)
    while improved:
        improved = False
        for i in range(len(best) - 1):
            for j in range(i + 1, len(best)):
                candidate = best[:i] + best[i : j + 1][::-1] + best[j + 1 :]
                cand_len = tour_len(candidate)
                if cand_len < best_len - 1e-9:
                    best, best_len = candidate, cand_len
                    improved = True
    return [picks[i] for i in best]


def pick_route(
    db: Session, org_id: int, warehouse_id: int,
    items: list | None, location_ids: list[int] | None,
) -> PickRouteOut:
    warehouse = get_owned_warehouse(db, org_id, warehouse_id)
    picks = _resolve_picks(db, org_id, warehouse_id, items, location_ids)

    depot = (warehouse.local_width / 2, 0.0)
    front_y = 0.0
    all_x = [p.x for p in picks]
    corridor_left = max(0.5, min(all_x) - 1.5)
    corridor_right = min(warehouse.local_width - 0.5, max(all_x) + 1.5)

    # Assert networkx presence for the optimized policy's contract (import
    # check keeps the dependency honest even though we hand-roll distances).
    assert nx is not None

    routes: list[PolicyRoute] = []
    orders = {
        "s_shape": _s_shape_order(picks),
        "largest_gap": _largest_gap_order(picks, corridor_left, corridor_right),
        "optimized": _optimized_order(depot, picks, front_y, corridor_left, corridor_right),
    }
    for policy, ordered in orders.items():
        total_m, path = _route_for_order(
            depot, ordered, front_y, corridor_left, corridor_right
        )
        routes.append(
            PolicyRoute(
                policy=policy,  # type: ignore[arg-type]
                total_m=total_m,
                stops=[
                    PickStop(
                        order=i + 1,
                        location_id=p.location_id,
                        code=p.code,
                        x=p.x,
                        y=p.y,
                        product_sku=p.product_sku,
                        quantity=p.quantity,
                    )
                    for i, p in enumerate(ordered)
                ],
                path=path,
            )
        )

    best = min(routes, key=lambda r: r.total_m)
    return PickRouteOut(
        warehouse_id=warehouse_id,
        pick_count=len(picks),
        routes=routes,
        best_policy=best.policy,
    )

File: frontend/src/demo/picking.ts
/** Demo pick-route solver: rectilinear walking over a front corridor,
 * mirroring the real backend's policy comparison (S-shape / largest-gap /
 * greedy+2-opt) at demo scale.
 */

import type { PickRoute, PickStop, PolicyRoute } from "@/types";
import { binQuantity, binStockRows, binsByWarehouse, warehouses } from "@/demo/data";

interface Pick {
  locationId: number;
  code: string;
  x: number; // bin center x
  y: number; // bin front y
  band: number;
}

const CORRIDOR_Y = 0.6;

function toPick(loc: { id: number; code: string; pos_x: number; pos_y: number; dim_w: number }): Pick {
  return {
    locationId: loc.id,
    code: loc.code,
    x: loc.pos_x + loc.dim_w / 2,
    y: loc.pos_y,
    band: Math.round(loc.pos_y * 10),
  };
}

/** Rectilinear path via the front corridor: walk x on the corridor, dip to the
 * bin, come back. Consecutive picks in the same band connect directly. */
function buildRoute(depot: { x: number; y: number }, order: Pick[]): { path: { x: number; y: number }[]; total: number } {
  const path: { x: number; y: number }[] = [{ x: depot.x, y: depot.y }];
  let cursor = { x: depot.x, y: depot.y, band: -1 };
  for (const pick of order) {
    if (cursor.band === pick.band) {
      path.push({ x: pick.x, y: pick.y });
    } else {
      if (cursor.y !== CORRIDOR_Y) path.push({ x: cursor.x, y: CORRIDOR_Y });
      path.push({ x: pick.x, y: CORRIDOR_Y });
      path.push({ x: pick.x, y: pick.y });
    }
    cursor = { x: pick.x, y: pick.y, band: pick.band };
  }
  if (cursor.y !== CORRIDOR_Y) path.push({ x: cursor.x, y: CORRIDOR_Y });
  path.push({ x: depot.x, y: depot.y });

  let total = 0;
  for (let i = 1; i < path.length; i++) {
    total += Math.abs(path[i].x - path[i - 1].x) + Math.abs(path[i].y - path[i - 1].y);
  }
  return { path, total };
}

function routeLength(depot: { x: number; y: number }, order: Pick[]): number {
  return buildRoute(depot, order).total;
}

function sShapeOrder(picks: Pick[]): Pick[] {
  const bands = [...new Set(picks.map((p) => p.band))].sort((a, b) => a - b);
  const out: Pick[] = [];
  bands.forEach((band, i) => {
    const inBand = picks.filter((p) => p.band === band).sort((a, b) => a.x - b.x);
    out.push(...(i % 2 === 0 ? inBand : inBand.reverse()));
  });
  return out;
}

function largestGapOrder(picks: Pick[]): Pick[] {
  const bands = [...new Set(picks.map((p) => p.band))].sort((a, b) => a - b);
  const out: Pick[] = [];
  for (const band of bands) {
    // enter/leave from the same end past the largest gap between picks
    const inBand = picks.filter((p) => p.band === band).sort((a, b) => a.x - b.x);
    let gapAt = 0;
    let gap = -1;
    for (let i = 1; i < inBand.length; i++) {
      const g = inBand[i].x - inBand[i - 1].x;
      if (g > gap) {
        gap = g;
        gapAt = i;
      }
    }
    out.push(...inBand.slice(0, gapAt), ...inBand.slice(gapAt).reverse());
  }
  return out;
}

function optimizedOrder(depot: { x: number; y: number }, picks: Pick[]): Pick[] {
  // greedy nearest neighbour on the corridor metric…
  const remaining = [...picks];
  const order: Pick[] = [];
  let cx = depot.x;
  while (remaining.length) {
    let best = 0;
    let bestD = Infinity;
    remaining.forEach((p, i) => {
      const d = Math.abs(p.x - cx) + Math.abs(p.y - CORRIDOR_Y);
      if (d < bestD) {
        bestD = d;
        best = i;
      }
    });
    const [next] = remaining.splice(best, 1);
    order.push(next);
    cx = next.x;
  }
  // …then hand-rolled 2-opt
  let improved = true;
  while (improved) {
    improved = false;
    for (let i = 0; i < order.length - 1; i++) {
      for (let j = i + 1; j < order.length; j++) {
        const candidate = [...order.slice(0, i), ...order.slice(i, j + 1).reverse(), ...order.slice(j + 1)];
        if (routeLength(depot, candidate) + 1e-9 < routeLength(depot, order)) {
          order.splice(0, order.length, ...candidate);
          improved = true;
        }
      }
    }
  }
  return order;
}

export function solvePickRoute(warehouseId: number, locationIds: number[]): PickRoute | { error: string } {
  const wh = warehouses.find((w) => w.id === warehouseId);
  const bins = binsByWarehouse.get(warehouseId) ?? [];
  const picks = locationIds
    .map((id) => bins.find((b) => b.id === id))
    .filter((b): b is NonNullable<typeof b> => Boolean(b))
    .map(toPick);
  if (!wh || picks.length < 2) {
    return { error: "Rota için bu depoda en az 2 geçerli göz gerekli." };
  }

  const depot = { x: wh.local_width / 2, y: 0.2 };
  const policies: [PolicyRoute["policy"], Pick[]][] = [
    ["s_shape", sShapeOrder(picks)],
    ["largest_gap", largestGapOrder(picks)],
    ["optimized", optimizedOrder(depot, picks)],
  ];

  const routes: PolicyRoute[] = policies.map(([policy, order]) => {
    const { path, total } = buildRoute(depot, order);
    const stops: PickStop[] = order.map((p, i) => {
      const row = binStockRows(p.locationId)[0];
      return {
        order: i + 1,
        location_id: p.locationId,
        code: p.code,
        x: p.x,
        y: p.y,
        product_sku: row?.sku ?? null,
        quantity: row ? Math.min(row.quantity, 5) : binQuantity(p.locationId) || null,
      };
    });
    return { policy, total_m: Math.round(total * 10) / 10, stops, path };
  });

  const best = routes.reduce((a, b) => (b.total_m < a.total_m ? b : a));
  return {
    warehouse_id: warehouseId,
    pick_count: picks.length,
    routes,
    best_policy: best.policy,
  };
}
